File: app/src/main/java/com/sayler/gina/activity/DayEditActivity.java
package com.sayler.gina.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.annimon.stream.Stream;
import com.sayler.gina.GinaApplication;
import com.sayler.gina.IAttachment;
import com.sayler.gina.IDay;
import com.sayler.gina.R;
import com.sayler.gina.interactor.days.ObjectCreator;
import com.sayler.gina.presenter.diary.DiaryPresenter;
import com.sayler.gina.presenter.diary.DiaryPresenterView;
import com.sayler.gina.util.Constants;
import com.sayler.gina.util.FileUtils;
import com.wdullaer.materialdatetimepicker.date.DatePickerDialog;
import icepick.Icepick;
import org.joda.time.DateTime;
import realm.DataManager;

import javax.inject.Inject;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class DayEditActivity extends BaseActivity implements DiaryPresenterView, DatePickerDialog.OnDateSetListener {

  private static final String TAG = "DayEditActivity";
  @Inject
  DiaryPresenter diaryPresenter;
  @Inject
  ObjectCreator objectCreator;
  @Inject
  DataManager dataManager;

  public Long dayId = -1L;
  public IDay day;

  @Bind(R.id.day)
  public TextView dayText;
  @Bind(R.id.year_month)
  public TextView yearMonthText;
  @Bind(R.id.content)
  public EditText contentText;
  @Bind(R.id.attachmentsContainer)
  public ViewGroup attachmentsContainer;

  private EditMode editMode;
  private AttachmentsManager attachmentsManager;

  private enum EditMode {
    NEW_DAY, EDIT_DAY

  }

  private class AttachmentsManager {
    private HashMap<IAttachment, Button> tmpAttachmentButtonHashMap = new HashMap<>();

    private ViewGroup attachmentsContainer;

    public AttachmentsManager(ViewGroup attachmentsContainer) {
      this.attachmentsContainer = attachmentsContainer;
    }

    public void addFile(byte[] bytes, String mimeType) {
      IAttachment newAttachment = objectCreator.createAttachment();
      newAttachment.setFile(bytes);
      newAttachment.setMimeType(mimeType);

      Button newButton = new Button(attachmentsContainer.getContext());
      newButton.setText(mimeType);
      newButton.setOnClickListener(view -> {
            tmpAttachmentButtonHashMap.remove(newAttachment);
            attachmentsContainer.removeView(view);
          }
      );

      attachmentsContainer.addView(newButton);
      tmpAttachmentButtonHashMap.put(newAttachment, newButton);
    }

    public List<IAttachment> returnAttachments() {
      List<IAttachment> attachments = new ArrayList<>();
      Stream.of(tmpAttachmentButtonHashMap).forEach(attachmentButtonEntry -> {
        attachments.add(attachmentButtonEntry.getKey());
      });

      return attachments;
    }

  }

  public static Intent newIntentEditDay(Context context, long dayId) {
    Intent intent = new Intent(context, DayEditActivity.class);
    intent.putExtra(Constants.EXTRA_DAY_ID, dayId);
    intent.putExtra(Constants.EXTRA_EDIT_MODE, EditMode.EDIT_DAY.ordinal());
    return intent;
  }

  public static Intent newIntentNewDay(Context context) {
    Intent intent = new Intent(context, DayEditActivity.class);
    intent.putExtra(Constants.EXTRA_EDIT_MODE, EditMode.NEW_DAY.ordinal());
    return intent;
  }

  public static void sendEditDayBroadcast(Context context) {
    Intent intent = new Intent(Constants.BROADCAST_EDIT_DAY);
    context.sendBroadcast(intent);
  }

  public static void sendDeleteDayBroadcast(Context context) {
    Intent intent = new Intent(Constants.BROADCAST_DELETE_DAY);
    context.sendBroadcast(intent);
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.a_edit_day);
    ButterKnife.bind(this);

    GinaApplication.getDataComponentForActivity(this).inject(this);

    attachmentsManager = new AttachmentsManager(attachmentsContainer);

    bindPresenters();

    readExtras();

    load();
  }

  @OnClick({R.id.year_month, R.id.day})
  public void onFabEditClick() {
    DatePickerDialog dpd = DatePickerDialog.newInstance(
        DayEditActivity.this,
        day.getDate().getYear(),
        day.getDate().getMonthOfYear() - 1,
        day.getDate().getDayOfMonth()
    );
    dpd.show(getFragmentManager(), DatePickerDialog.class.getCanonicalName());
  }

  private void readExtras() {
    if (getIntent().hasExtra(Constants.EXTRA_EDIT_MODE)) {
      //get edit mode
      int editModeOrdinal = getIntent().getExtras().getInt(Constants.EXTRA_EDIT_MODE);
      editMode = EditMode.values()[editModeOrdinal];

      switch (editMode) {
        case NEW_DAY:
          day = objectCreator.createDay();
          day.setDate(new DateTime());
          break;
        case EDIT_DAY:
          dayId = getIntent().getLongExtra(Constants.EXTRA_DAY_ID, -1);
          break;
      }
    }
  }

  private void bindPresenters() {
    diaryPresenter.onBindView(this);
  }

  private void load() {
    switch (editMode) {
      case NEW_DAY:
        showTextContent();
        break;
      case EDIT_DAY:
        diaryPresenter.loadById(dayId);
        break;
    }
  }

  @OnClick(R.id.fab_save)
  public void onFabSaveClick() {
    put();
  }

  @OnClick(R.id.fab_delete)
  public void onFabDeleteClick() {
    delete();
  }

  @OnClick(R.id.fab_add_attachment)
  public void onFabAddAttachmentClick() {
    FileUtils.selectFileIntent(this, Constants.REQUEST_CODE_SELECT_ATTACHMENT);
  }

  @Override
  protected void onActivityResult(int requestCode, int resultCode, Intent data) {
    super.onActivityResult(requestCode, resultCode, data);
    if (requestCode == Constants.REQUEST_CODE_SELECT_ATTACHMENT) {
      addAttachment(data);
    }
  }

  private void put() {
    day.setContent(contentText.getText().toString());

    diaryPresenter.put(day, attachmentsManager.returnAttachments());
  }

  private void delete() {
    diaryPresenter.delete(day);
  }

  @Override
  public void onPut() {
    sendEditDayBroadcast(this);
    dataManager.close();
    finish();
  }

  @Override
  public void onDownloaded(List<IDay> data) {
    day = data.get(0);
    showTextContent();
    showAttachments();
  }

  @Override
  public void onDelete() {
    sendDeleteDayBroadcast(this);
    dataManager.close();
    finish();
  }

  @Override
  public void onNoDataSource() {
    //TODO error handling
  }

  @Override
  public void onError(String errorMessage) {
    showError(errorMessage);
  }

  private void showError(String errorMessage) {
    Snackbar.make(findViewById(android.R.id.content), errorMessage, Snackbar.LENGTH_SHORT).show();
  }

  private void showTextContent() {
    dayText.setText(day.getDate().toString(Constants.DATA_PATTERN_DAY_NUMBER_DAY_OF_WEEK));
    yearMonthText.setText(day.getDate().toString(Constants.DATE_PATTERN_YEAR_MONTH));
    contentText.setText(day.getContent());
  }

  private void showAttachments() {
    for (IAttachment attachment : day.getAttachments()) {
      attachmentsManager.addFile(attachment.getFile(), attachment.getMimeType());
    }
  }

  private void addAttachment(Intent data) {
    try {
      byte[] fileBytes = FileUtils.readFileFromUri(data.getData(), this);
      String mimeType = FileUtils.readMimeTypeFromUri(data.getData(), this);

      attachmentsManager.addFile(fileBytes, mimeType);
    } catch (IOException e) {
      e.printStackTrace();
      //TODO error handling
    }
  }

  @Override
  protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    Icepick.saveInstanceState(this, outState);
  }

  @Override
  protected void onRestoreInstanceState(Bundle savedInstanceState) {
    super.onRestoreInstanceState(savedInstanceState);
    Icepick.restoreInstanceState(this, savedInstanceState);
  }

  @Override
  protected void onDestroy() {
    diaryPresenter.onUnBindView();
    dataManager.close();
    super.onDestroy();
  }

  @Override
  public void onDateSet(DatePickerDialog view, int year, int monthOfYear, int dayOfMonth) {
    DateTime dateTime = new DateTime(year, monthOfYear + 1, dayOfMonth, 0, 0);
    day.setDate(dateTime);
    day.setContent(contentText.getText().toString());
    showTextContent();
  }
}

File: app/src/main/java/com/sayler/gina/activity/DayEditActivity.kt
package com.sayler.gina.activity

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v4.view.GravityCompat
import android.support.v7.widget.LinearLayoutManager
import android.view.Gravity
import butterknife.ButterKnife
import butterknife.OnClick
import butterknife.OnTouch
import com.jakewharton.rxbinding2.widget.textChanges
import com.sayler.gina.GinaApplication
import com.sayler.gina.R
import com.sayler.gina.attachment.AttachmentAdapter
import com.sayler.gina.attachment.AttachmentManagerContract
import com.sayler.gina.domain.DataManager
import com.sayler.gina.domain.IAttachment
import com.sayler.gina.domain.IDay
import com.sayler.gina.domain.ObjectCreator
import com.sayler.gina.domain.presenter.diary.DiaryContract
import com.sayler.gina.util.AlertUtility
import com.sayler.gina.util.Constants
import com.sayler.gina.util.FileUtils
import com.wdullaer.materialdatetimepicker.date.DatePickerDialog
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.a_day_edit.*
import org.joda.time.DateTime
import java.io.IOException
import java.util.concurrent.TimeUnit
import javax.inject.Inject

class DayEditActivity : BaseActivity(), DatePickerDialog.OnDateSetListener {
    @Inject
    lateinit var diaryPresenter: DiaryContract.Presenter
    @Inject
    lateinit var objectCreator: ObjectCreator
    @Inject
    lateinit var dataManager: DataManager<*>
    @Inject
    lateinit var attachmentManager: AttachmentManagerContract.Presenter

    private lateinit var day: IDay
    private lateinit var editMode: EditMode
    private var dayId: Long = -1L
    private var initialHashDay: Int = -1
    private var initialHashAttachments: Int = -1
    private lateinit var fabs: List<FloatingActionButton>
    private lateinit var attachmentAdapter: AttachmentAdapter

    private enum class EditMode {
        NEW_DAY, EDIT_DAY
    }

    private val diaryContractView = object : DiaryContract.View {
        override fun showProgress() {
            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        }

        override fun hideProgress() {
            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        }

        override fun noDataSource() {
            Snackbar.make(findViewById(android.R.id.content), R.string.no_data_source, Snackbar.LENGTH_SHORT).show()
        }

        override fun onPut() {
            sendEditDayBroadcast(this@DayEditActivity)
            dataManager.close()
            finish()
        }

        override fun onDownloaded(data: List<IDay>) {
            day = data[0]
            initialHashDay = day.hashCode()
            initialHashAttachments = day.attachments.hashCode()
            setupDay()
        }

        override fun onDelete() {
            sendDeleteDayBroadcast(this@DayEditActivity)
            dataManager.close()
            finish()
        }

        override fun onError(errorMessage: String) {
            showError(errorMessage)
        }
    }

    private val attachmentManagerView = object : AttachmentManagerContract.View {
        override fun showProgress() {
            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        }

        override fun hideProgress() {
            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        }

        override fun noDataSource() {
            TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
        }

        override fun onUpdate(attachments: MutableCollection<IAttachment>) {
            attachmentAdapter.updateItems(attachments)
            attachmentAdapter.notifyDataSetChanged()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.a_day_edit)

        fabs = listOf(fab_add_attachment, fab_delete, fab_save, fabAttachments)

        ButterKnife.bind(this)
        GinaApplication.dataComponentForActivity(this).inject(this)

        bindPresenters()

        readExtras()

        setupViews()

        load()
    }

    private fun setupViews() {
        content.textChanges()
                .skip(2)
                .map { day.content = content.text.toString() }
                .doOnNext { fabs.onEach { it.hide() } }
                .buffer(2, TimeUnit.SECONDS)
                .filter { t -> t.size <= 0 }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { fabs.onEach { it.show() } }
    }

    private fun readExtras() {
        if (intent.hasExtra(Constants.EXTRA_EDIT_MODE)) {
            //get edit mode
            val editModeOrdinal = intent.extras.getInt(Constants.EXTRA_EDIT_MODE)
            editMode = EditMode.values()[editModeOrdinal]

            when (editMode) {
                DayEditActivity.EditMode.NEW_DAY -> {
                    day = objectCreator.createDay()
                    day.date = DateTime()
                    initialHashDay = day.hashCode()
                    initialHashAttachments = day.attachments.hashCode()
                }
                DayEditActivity.EditMode.EDIT_DAY -> dayId = intent.getLongExtra(Constants.EXTRA_DAY_ID, -1)
            }
        } else {
            throw IllegalStateException("Missing intent extra params: EXTRA_EDIT_MODE")
        }
    }

    private fun bindPresenters() {
        bindPresenter(attachmentManager, attachmentManagerView)
        bindPresenter(diaryPresenter, diaryContractView)
    }


    private fun load() {
        when (editMode) {
            DayEditActivity.EditMode.NEW_DAY -> setupDay()
            DayEditActivity.EditMode.EDIT_DAY -> diaryPresenter.loadById(dayId)
        }
    }

    private fun setupDay() {
        showTextContent()
        setupAttachments()
    }

    private fun showTextContent() {
        dayText.text = day.date.toString(Constants.DATA_PATTERN_DAY_NUMBER_DAY_OF_WEEK)
        yearMonthText.text = day.date.toString(Constants.DATE_PATTERN_YEAR_MONTH)
        content.setText(day.content)
        if (content.text.isNotEmpty())
            content.setSelection(content.text.length - 1)
    }

    private fun setupAttachments() {
        //setup drawer with attachment
        val layoutManager = LinearLayoutManager(this)
        attachmentsRecyclerView.layoutManager = layoutManager
        attachmentAdapter = AttachmentAdapter(day.attachments, attachmentsRecyclerView, true)
        //on item click
        attachmentAdapter.setOnClick({ item, _ ->
            with(item.attachment) {
                FileUtils.openFileIntent(this@DayEditActivity, file, mimeType, applicationContext.packageName + ".provider")
            }
        })
        //on remove button click on item
        attachmentAdapter.setOnRemoveClick { item ->
            AlertUtility.showConfirmationAlert(this, R.string.remove_attachment_title, R.string.remove_attachment_confirmation, DialogInterface.OnClickListener { dialogInterface, i ->
                attachmentManager.remove(item.attachment)
            })
        }
        attachmentsRecyclerView.adapter = attachmentAdapter

        //setup attachmentManager
        attachmentManager.setup(day.attachments)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode != RESULT_CANCELED && requestCode == Constants.REQUEST_CODE_SELECT_ATTACHMENT && data != null) {
            //multiple files
            if (data.clipData != null) {
                for (i in 0..data.clipData.itemCount - 1)
                    addAttachment(data.clipData.getItemAt(i).uri)
            } else {//single file
                addAttachment(data.data)
            }
        }
    }

    @OnTouch(R.id.root, R.id.content)
    fun onElseClick(): Boolean {
        fabs.onEach { it.show() }
        return false
    }

    @OnClick(R.id.yearMonthText, R.id.dayText)
    fun onFabEditClick() {
        val dpd = DatePickerDialog.newInstance(
                this@DayEditActivity,
                day.date.year,
                day.date.monthOfYear - 1,
                day.date.dayOfMonth
        )
        dpd.show(fragmentManager, DatePickerDialog::class.java.canonicalName)
    }

    override fun onBackPressed() {
        if (drawer_layout.isDrawerOpen(Gravity.END)) {
            drawer_layout.closeDrawer(Gravity.END)
            return
        }

        if (hasDayChanged()) {
            AlertUtility.showConfirmationAlert(this, R.string.discard_title, R.string.discard_confirmation,
                    DialogInterface.OnClickListener { _, _ -> super.onBackPressed() })
            return
        }
        super.onBackPressed()
    }

    private fun hasDayChanged() =
            initialHashDay != day.hashCode() || initialHashAttachments != attachmentManager.getAll().toList().hashCode()

    @OnClick(R.id.fabAttachments)
    fun onFabAttachmentsClick() {
        drawer_layout.openDrawer(GravityCompat.END)
    }

    @OnClick(R.id.fab_save)
    fun onFabSaveClick() {
        put()
    }

    @OnClick(R.id.fab_delete)
    fun onFabDeleteClick() {
        AlertUtility.showConfirmationAlert(this, R.string.remove_day_title, R.string.remove_day_confirmation, DialogInterface.OnClickListener { dialogInterface, i ->
            delete()
        })
    }

    @OnClick(R.id.fab_add_attachment)
    fun onFabAddAttachmentClick() {
        FileUtils.selectFileIntent(this, Constants.REQUEST_CODE_SELECT_ATTACHMENT)
    }

    private fun put() {
        diaryPresenter.put(day, attachmentManager.getAll().toList())
    }

    private fun delete() {
        diaryPresenter.delete(day)
    }

    private fun showError(errorMessage: String) {
        Snackbar.make(findViewById(android.R.id.content), errorMessage, Snackbar.LENGTH_SHORT).show()
    }

    private fun addAttachment(uri: Uri) {
        try {
            val fileBytes = FileUtils.readFileFromUri(uri, this)
            val mimeType = FileUtils.readMimeTypeFromUri(uri, this)

            attachmentManager.add(fileBytes, mimeType)
        } catch (e: IOException) {
            e.printStackTrace()
            showError(getString(R.string.attachment_add_error))
        }

    }

    override fun onDestroy() {
        dataManager.close()
        super.onDestroy()
    }

    override fun onDateSet(view: DatePickerDialog, year: Int, monthOfYear: Int, dayOfMonth: Int) {
        val dateTime = DateTime(year, monthOfYear + 1, dayOfMonth, 0, 0)
        day.date = dateTime
        day.content = content.text.toString()
        showTextContent()
    }

    companion object {

        fun newIntentEditDay(context: Context, dayId: Long): Intent {
            val intent = Intent(context, DayEditActivity::class.java)
            intent.putExtra(Constants.EXTRA_DAY_ID, dayId)
            intent.putExtra(Constants.EXTRA_EDIT_MODE, EditMode.EDIT_DAY.ordinal)
            return intent
        }

        fun newIntentNewDay(context: Context): Intent {
            val intent = Intent(context, DayEditActivity::class.java)
            intent.putExtra(Constants.EXTRA_EDIT_MODE, EditMode.NEW_DAY.ordinal)
            return intent
        }

        fun sendEditDayBroadcast(context: Context) {
            val intent = Intent(Constants.BROADCAST_EDIT_DAY)
            context.sendBroadcast(intent)
        }

        fun sendDeleteDayBroadcast(context: Context) {
            val intent = Intent(Constants.BROADCAST_DELETE_DAY)
            context.sendBroadcast(intent)
        }
    }
}
